staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/validator.go

```diff
@@ -101,30 +106,57 @@ func (v *validator) Validate(ctx context.Context, versionedAttr *admission.Versi
 			continue
 		}
 
+		var messageResult *cel.EvaluationResult
+		var messageError *apiservercel.Error
+		if len(messageResults) > i {
+			messageResult = &messageResults[i]
+		}
+		messageError, _ = err.(*apiservercel.Error)
 		if evalResult.Error != nil {
 			decision.Action = policyDecisionActionForError(f)
 			decision.Evaluation = EvalError
 			decision.Message = evalResult.Error.Error()
+		} else if messageError != nil &&
+			(messageError.Type == apiservercel.ErrorTypeInternal ||
+				(messageError.Type == apiservercel.ErrorTypeInvalid &&
+					strings.HasPrefix(messageError.Detail, "validation failed due to running out of cost budget"))) {
+			decision.Action = policyDecisionActionForError(f)
+			decision.Evaluation = EvalError
+			decision.Message = fmt.Sprintf("failed messageExpression: %s", err)
 		} else if evalResult.EvalResult != celtypes.True {
 			decision.Action = ActionDeny
 			if validation.Reason == nil {
 				decision.Reason = metav1.StatusReasonInvalid
 			} else {
 				decision.Reason = *validation.Reason
 			}
-			if len(validation.Message) > 0 {
-				decision.Message = strings.TrimSpace(validation.Message)
-			} else {
-				decision.Message = fmt.Sprintf("failed expression: %v", strings.TrimSpace(validation.Expression))
+			// decide the failure message
+			var message string
+			// attempt to set message with messageExpression result
+			if messageResult != nil && messageResult.Error == nil && messageResult.EvalResult != nil {
+				// also fallback if the eval result is non-string (including null) or
+				// whitespaces.
+				if message, ok = messageResult.EvalResult.Value().(string); ok {
+					message = strings.TrimSpace(message)
+				}
+			}
+			// fallback to set message to the custom message
+			if message == "" && len(validation.Message) > 0 {
+				message = strings.TrimSpace(validation.Message)
+			}
+			// fallback to use the expression to compose a message
+			if message == "" {
+				message = fmt.Sprintf("failed expression: %v", strings.TrimSpace(validation.Expression))
 			}
+			decision.Message = message
 		} else {
 			decision.Action = ActionAdmit
 			decision.Evaluation = EvalAdmit
 		}
 	}
 
 	options := cel.OptionalVariableBindings{VersionedParams: versionedParams}
-	auditAnnotationEvalResults, err := v.auditAnnotationFilter.ForInput(ctx, versionedAttr, cel.CreateAdmissionRequest(versionedAttr.Attributes), options, runtimeCELCostBudget)
+	auditAnnotationEvalResults, _, err := v.auditAnnotationFilter.ForInput(ctx, versionedAttr, cel.CreateAdmissionRequest(versionedAttr.Attributes), options, runtimeCELCostBudget)
```

Comment:
I realized when reading this that I didn't share the budget between `validations` and `auditAnnoations`.  BTH, I didn't think it through carefully.  Any thoughts on if they should be kept separate or combined?  I'm currently in favor of keeping them separate (the way the code exists now) since `auditAnnotations` are logically their own capability separate from validation, but I'm curious what others think.

Reply:
@cici37 We had a discussion on this topic. Since it was a bit rushed we did not have a good idea about sharing cost vs dedicated cost. Could you help me answer this concern? Thank you.

Reply:
After offline discussion with @jpbetz , let's have the runtimeCELCostBudget only for validations which is shared by expression and messageExpression. auditAnnotations and possible matchConditions will have separate budgets.

Reply:
This behavior is consistent with current implementation.